XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+the `init()`, etc., calls from firing for any new roots of asynchronous call
+stacks, but will not prevent existing asynchronous call stacks that have
+already been captured by the `AsyncHook` instance from continuing to fire.
+
+Be careful to always disable the hooks when they are no longer needed.
+Especially in cases where there are multiple return statements. If the hooks
+are not disabled then they will stay active indefinitely. The following is an
+example of how a call to `disable()` may accidentally left out:
+
+```js
+const hooks = new async_wrap.AsyncHook(/* ... */);
+
+crypto.randomBytes(1024, (err, data) => {
+  hooks.enable();
+  try {
+    fs.writeFileSync(path, data);
+  } catch (e) {
+    console.error('file write failed');
+    // Notice the early return and failure to run hooks.disable(), which will
+    // allow the hooks callbacks to continue firing indefinitely.
+    return;
+  }
+  hooks.disable();
+});
+```
+
+While not included as part of initial development, it should be possible in the
+future to allow selective tracking of asynchronous call stacks. The following
+example demonstrates this:
+
+```js
+const async_wrap = require('async_wrap');
+const net = require('net');
+
+// Pretend init, before, after are all defined
+const asyncHook = async_wrap.createHook({ init, before, after });
+asyncHook.enable();
+
+net.createServer((c) => {
+  // Only want to follow connections that match IP range.
+  if (ipRangeRegExp.test(c.address().address))
+    asyncHook.disable();
+}).listen(PORT);
+```
+
+At which point no further calls to the hooks on that instance will be made for
+that asynchronous branch.
+
+
+#### `asyncHook.syncScope()`
+
+Enable capture of asynchronous events until the current synchronous code
+execution has completed. This is basically a small amount of sugar to prevent
+accidentally forgetting to `disable()` a set of hooks in a function with
+multiple returns. Also in case an error is thrown and caught by a domain or
+`uncaughtException`.
+
+Say an application only wishes to observe the asynchronous calls made from
+another asynchronous callback. Without `syncScope()` the call would look like
+so:
+
+```js
+fs.readFile(path, (err, data) => {
+  asyncHook.enable();
+  var e = asyncDataProcessing(data);
+  if (e) {
+    asyncHook.disable();
+    return;
+  }
+  doMoreAsyncStuff();
+  asyncHook.disable();
+});
+```
+
+But using `syncScope()` all of that is handled for you:
+
+```js
+fs.readFile(path, (err, data) => {
+  asyncHook.syncScope();
+  var e = asyncDataProcessing(data);
+  if (e)
+    return;
+  doMoreAsyncStuff();
+});
+```
+
+**Note**: Notice that this specifically states the synchronous scope. Not the
+function scope. So if this is used several call stacks deep keep in mind that
+the hooks will be enabled until the stack completely unwinds.
+
+
+#### `asyncHook.remove()`
+
+Unlike `disable()`, `remove()` prevents the callbacks from firing again.
```

Comment:
It will be possible to call `enable()` again. When it is called, it will act as if it's the first time `enable()` had been called. Maybe `clear()` would be a better term? Guess then I'd expect users would need to do `asyncHook.disable().clear()`. Thoughts?